Source/Core/VideoCommon/TmemBase.cpp

```diff
@@ -0,0 +1,203 @@
+// Copyright 2019 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include <array>
+
+#include "VideoCommon/BPMemory.h"
+#include "VideoCommon/TmemBase.h"
+
+namespace TmemBase
+{
+// On TMEM configuration changed:
+// 1. invalidate stage.
+
+struct texture_unit_state
+{
+  enum State
+  {
+    INVALID, // Configuration has changed
+    VALID, // But not cached due to either being too big, or overlapping with another valid texture unit
+    CACHED,
+  };
+
+  TexImage1 even;
+  TexImage2 odd;
+  State state;
+
+  int even_base;
+  int even_size;
+  int odd_base;
+  int odd_size;
+};
+static std::array<texture_unit_state, 8> s_unit;
+
+void ConfigurationChanged(int bp_addr, int config)
+{
+  // Extract bits encoding texture unit
+  u8 unit = ((bp_addr & 0x20) >> 3) | (bp_addr & 3);
+  bool even = bp_addr < BPMEM_TX_SETIMAGE2;
+
+  texture_unit_state& unit_state = s_unit[unit];
+
+  if (even)
+  {
+    unit_state.even.hex = config;
+  }
+  else
+  {
+    unit_state.odd.hex = config;
+  }
+  unit_state.state = texture_unit_state::INVALID;
+}
+
+// On invalidate cache
+// 1. invalidate all
+
+void Invalidate(u32 param)
+{
+  // The excat arguments of Invalidate commands is currently unknown.
+  // It appears to contain the TMEM address and a size.
+
+  // For simplicity, we will just invalidate everything
+  for (auto& unit : s_unit)
+  {
+    unit.state = texture_unit_state::INVALID;
+  }
+}
+
+// On bind:
+// 1. store the memory region(s) bound
+// 2. if texture size is small enough to fit in region AND there is no overlap, mark as cached.
+// 3. If there is cached overlap, invalidate.
+
+static inline u32 calculate_unit_size(TexImage1 config)
```

Comment:
`inline` is superfluous here given static is already specified. This should also be named `CalculateUnitSize`